contributors/design-proposals/sig-cli/apply_v2.md

```diff
@@ -0,0 +1,201 @@
+    beforehand)
+  - Patch hides values on the remote object, which may have hidden interactions with the patch values
+    (e.g. Deployment.Spec.RolloutStrategy)
+- Generating a patch to make a change is more complicated than simply making the change to
+  the object read from the server.  The apply command reads the live object from the server
+  and diffs it against the local configuration.  Building a patch containing special merge
+  directives is a more complicated approach than modifying the object read from the server
+  and updating it with a PUT.
+
+#### New approach
+
+- Read the live object
+- Compare the live object to last-recorded and local files
+- Update the fields on the live object that was read
+- Send a PUT to update the modified object
+- If encountering optimistic lock failure, retry after re-reading the object
+
+#### Trade offs
+
+Good:
+
+- Simpler code to maintain
+- Behaves correctly in more instances
+- Easy to look at the output and determine the final state
+
+Bad:
+
+- Larger request sizes to the server than when sending a patch
+
+### Restructure code into modular components
+
+In the current implementation of apply - parsing and traversing the object trees, diffing the
+contents and generating the patch are logically entangled.  This creates maintenance and
+testing challenges.  We should instead modularize discrete responsibilities - collating the
+object values and updating the target object.
+
+#### Phase 1: Parse last-recorded, local, live objects and collate
+
+Provide a structure that contains the last, local and live value for each field.  This
+will make it easy to walk the a single tree when making decisions about how to update the object.
+
+#### Phase 2: Diff and update objects
+
+Use the visitor pattern to encapsulate how to update each field type for each merge strategy.
+Unit test each visit function.
+
+## Alternatives considered
+
+Continue to use patch for apply.  Continue to evolve the PATCH semantics to be more
+forgiving of version skewed environments.  Add new directives as requirements identified.
+
+## Detailed analysis of structure and impact today
+
+The following PRs constitute the focus of ~6 months of engineering work.  Each of the PRs is very complex
+for the work what it is solving.
+
+### Patterns observed
+
+- PRs frequently closed or deferred because maintainers / reviewers cannot reason about the impact or
+  correctness of the changes
+- Relatively simple changes
+  - are 200+ lines of code
+  - modify dozens of existing locations in the code
+  - are spread across 1000+ lines of existing code
+- Changes that add new directives require updates in multiple locations - create patch + apply patch
+
+### PRs
+
+[38665](https://github.com/kubernetes/kubernetes/pull/38665/files)
+- Support deletion of primitives from lists
+- Lines (non-test): ~200
+- ~6 weeks
+[44597](https://github.com/kubernetes/kubernetes/pull/44597/files)
```

Comment:
This was the wrong solution to this problem, IMO.

Reply:
I want to understand why.

Reply:
We want the correct behavior regardless whether POST, PUT, PATCH (any flavor), or apply is used. Therefore, the proper fix would be in the API. My proposal in https://github.com/kubernetes/kubernetes/issues/35345 was to add discriminators. It was incorrectly concluded that "We should set discriminators by default", and was dismissed as unsafe without further explanation.

Regardless, unions/OneOf are a property of the API schema. A patch directive should not have been needed. The patch version could be specified via Content-Type, which is how we currently distinguish among the flavors of patches:
https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#patch-operations

Reply:
And it's not clear to me that changing default API behavior is any more or less a breaking change than changing default kubectl behavior with no way to preserve the previous behavior.

Reply:
And what would break, actually? A patch to another key in a union should result in validation failure in the case that the previously set key were not cleared. Making such a request succeed would not be a breaking change -- all requests that previously worked would still work the same way.

Reply:
I would be happy to see Union get implemented.  It will need a champion for that to happen.  As things stand today, don't see my team tackling this in the next year.

Reply:
Even without discriminated unions, I don't see why directives were necessary.